kernel/init.c

```diff
@@ -337,6 +340,12 @@ static int do_device_init(const struct device *dev)
 	return rc;
 }
 
+static inline bool is_entry_about_service(const void *obj)
+{
+	return (obj >= (void *)_service_list_start &&
+		obj < (void *)_service_list_end);
```

Comment:
This per-item test seems needless.  You're sorting them in the linker already as a "sub-section" of an existing contiguous level.  No need to check every item, just walk the prefix that is "devices" and the suffix which is "services", no?

Reply:
I am afraid I don't see what you are seeing. Service and device presence in init entries is shuffled, so one after another an init entry might point to a service or a device. What am I missing?